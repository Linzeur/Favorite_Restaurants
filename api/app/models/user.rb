class User < ApplicationRecord

  validates :name, presence: true
  validates :email, format: { with: /[a-zA-Z0-9]*@[a-zA-Z]*.[a-zA-Z0-9]*/, message: "Bad format"  },
                    uniqueness: true

  has_secure_password
  has_secure_token
  has_secure_token :reset_digest

  has_many :locations
  has_many :favorite_restaurants

  def self.valid_login?(email, password)
    user = find_by(email: email)
    user if user&.authenticate(password)
  end

  def invalidate_token
    update(token: nil)
  end
end
