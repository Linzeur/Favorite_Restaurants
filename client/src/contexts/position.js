import React from "react";

const PositionContext = React.createContext();

function PositionProvider(props) {
  const [position, setPosition] = React.useState([0, 0, false]);

  React.useEffect(() => {
    const watchId = navigator.geolocation.watchPosition(({ coords }) => {
      setPosition([coords.latitude, coords.longitude, true]);
      sessionStorage.setItem(
        "lastCurrentPosition",
        JSON.stringify({ lat: coords.latitude, lng: coords.longitude })
      );
    });
    return () => navigator.geolocation.clearWatch(watchId);
  }, [setPosition]);

  const value = {
    latitude: position[0],
    longitude: position[1],
    callAPI: position[2],
    onChange: (lat, lng, callAPI) => {
      setPosition([lat, lng, callAPI]);
    }
  };

  return <PositionContext.Provider value={value} {...props} />;
}

export { PositionProvider, PositionContext };
