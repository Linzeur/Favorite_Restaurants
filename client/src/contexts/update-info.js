import React from "react";

const UpdateListContext = React.createContext();

function UpdateListProvider(props) {
  const [updateList, setUpdateList] = React.useState({
    locations: 0,
    favorites: 0
  });

  const value = {
    locations: updateList.locations,
    favorites: updateList.favorites,
    onChange: key => {
      setUpdateList(prevState => {
        return { ...prevState, [key]: prevState[key] + 1 };
      });
    }
  };

  return <UpdateListContext.Provider value={value} {...props} />;
}

export { UpdateListProvider, UpdateListContext };
