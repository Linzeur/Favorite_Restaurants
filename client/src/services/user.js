import { apiUrl } from "../util";

const API_USERS = `${apiUrl}users`;

async function createUser(userData) {
  const response = await fetch(API_USERS, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(userData)
  });

  if (!response.ok) {
    const { errors } = await response.json();
    throw new Error(errors.message);
  }

  return response.json();
}

export { createUser };
