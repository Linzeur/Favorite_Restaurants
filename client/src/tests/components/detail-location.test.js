import React from "react";
import { render, fireEvent } from "@testing-library/react";

import DetailLocation from "../../components/detail-location";
import { UserProvider } from "../../contexts/user";
import { UpdateListProvider } from "../../contexts/update-info";

test("detail-location component to create new location", () => {
  const setModalFn = jest.fn();

  const { getByLabelText, getByText, asFragment } = render(
    <UserProvider>
      <UpdateListProvider>
        <DetailLocation showModalFn={setModalFn} />
      </UpdateListProvider>
    </UserProvider>
  );

  expect(asFragment()).toMatchSnapshot();
  const modalEditLocation = getByText("New Location");
  const modalInputName = getByLabelText("Enter Name");
  const modalInputAddress = getByLabelText("Enter Address");
  const modalInputDistrict = getByLabelText("Enter District");

  expect(modalEditLocation).not.toBeNull();
  expect(modalInputName.value).toBe("");
  expect(modalInputAddress.value).toBe("");
  expect(modalInputDistrict.value).toBe("");
});

const location = {
  id: 1,
  name: "Ovalo Miraflores",
  address: "Parque Kennedy",
  district: "Miraflores"
};

test("detail-location component to edit an existing location", async () => {
  const setModalFn = jest.fn();

  const { getByLabelText, getByText, asFragment } = render(
    <UserProvider>
      <UpdateListProvider>
        <DetailLocation locationDetail={location} showModalFn={setModalFn} />
      </UpdateListProvider>
    </UserProvider>
  );

  expect(asFragment()).toMatchSnapshot();

  const modalEditLocation = getByText("Edit Location");
  const modalInputName = getByLabelText("Enter Name");
  const modalInputAddress = getByLabelText("Enter Address");
  const modalInputDistrict = getByLabelText("Enter District");

  expect(modalEditLocation).not.toBeNull();
  expect(modalInputName.value).toBe(location.name);
  expect(modalInputAddress.value).toBe(location.address);
  expect(modalInputDistrict.value).toBe(location.district);
});

test("click in button close", async () => {
  const setModalFn = jest.fn();

  const { getAllByRole, asFragment } = render(
    <UserProvider>
      <UpdateListProvider>
        <DetailLocation locationDetail={location} showModalFn={setModalFn} />
      </UpdateListProvider>
    </UserProvider>
  );

  const buttonClose = getAllByRole("button")[0];

  fireEvent.click(buttonClose);

  expect(setModalFn).toHaveBeenCalledWith(null);
});
