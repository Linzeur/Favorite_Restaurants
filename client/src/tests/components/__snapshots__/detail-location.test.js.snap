// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail-location component to create new location 1`] = `
<DocumentFragment>
  <section
    class="css-wh9mvb-DetailLocation"
  >
    <form
      class="css-1e3smbc-DetailLocation"
    >
      <div
        class="css-r4e2nb-DetailLocation"
      >
        <h2>
          New Location
        </h2>
      </div>
      <hr />
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Name
        </label>
        <input
          aria-label="Enter Name"
          autocomplete="off"
          class="css-1yxseax"
          id="Name"
          name="Name"
          placeholder="Enter Name"
          required=""
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Address
        </label>
        <input
          aria-label="Enter Address"
          autocomplete="off"
          class="css-1yxseax"
          id="Address"
          name="Address"
          placeholder="Enter Address"
          required=""
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          District
        </label>
        <input
          aria-label="Enter District"
          autocomplete="off"
          class="css-1yxseax"
          id="District"
          name="District"
          placeholder="Enter District"
          required=""
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        class="css-qm5crd-DetailLocation"
      >
        <button
          class="css-1sujpum-DetailLocation"
          type="button"
        >
          Cancel
        </button>
        <button
          class="css-1sujpum-DetailLocation"
        >
          Save
        </button>
      </fieldset>
    </form>
  </section>
</DocumentFragment>
`;

exports[`detail-location component to edit an existing location 1`] = `
<DocumentFragment>
  <section
    class="css-wh9mvb-DetailLocation"
  >
    <form
      class="css-1e3smbc-DetailLocation"
    >
      <div
        class="css-r4e2nb-DetailLocation"
      >
        <h2>
          Edit Location
        </h2>
      </div>
      <hr />
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Name
        </label>
        <input
          aria-label="Enter Name"
          autocomplete="off"
          class="css-1yxseax"
          id="Name"
          name="Name"
          placeholder="Enter Name"
          required=""
          type="text"
          value="Ovalo Miraflores"
        />
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          Address
        </label>
        <input
          aria-label="Enter Address"
          autocomplete="off"
          class="css-1yxseax"
          id="Address"
          name="Address"
          placeholder="Enter Address"
          required=""
          type="text"
          value="Parque Kennedy"
        />
      </fieldset>
      <fieldset
        class="css-1ihc67e"
      >
        <label
          class="css-19qo4af"
        >
          District
        </label>
        <input
          aria-label="Enter District"
          autocomplete="off"
          class="css-1yxseax"
          id="District"
          name="District"
          placeholder="Enter District"
          required=""
          type="text"
          value="Miraflores"
        />
      </fieldset>
      <fieldset
        class="css-qm5crd-DetailLocation"
      >
        <button
          class="css-1sujpum-DetailLocation"
          type="button"
        >
          Cancel
        </button>
        <button
          class="css-1sujpum-DetailLocation"
        >
          Save
        </button>
      </fieldset>
    </form>
  </section>
</DocumentFragment>
`;
